CefSharp/Internals/MethodRunnerQueue.cs

```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CefSharp.Internals
+{
+    public sealed class MethodRunnerQueue
+    {
+        private readonly JavascriptObjectRepository repository;
+        private readonly AutoResetEvent stopped = new AutoResetEvent(false);
+        private readonly BlockingCollection<Task<MethodInvocationResult>> queue = new BlockingCollection<Task<MethodInvocationResult>>();
+        private readonly object lockObject = new object();
+        private volatile CancellationTokenSource cancellationTokenSource;
+        private volatile bool running;
+
+        public event EventHandler<MethodInvocationCompleteArgs> MethodInvocationComplete;
+
+        public MethodRunnerQueue(JavascriptObjectRepository repository)
+        {
+            this.repository = repository;
+        }
+
+        public void Start()
+        {
+            if (running)
+                return;
+
+            lock (lockObject)
+            {
+                if (!running)
+                {
+                    cancellationTokenSource = new CancellationTokenSource();
+                    Task.Factory.StartNew(ConsumeTasks, TaskCreationOptions.LongRunning);
+                    running = true;
+                }
+            }
+        }
+
+        public void Stop()
+        {
+            if (!running)
+                return;
+
+            lock (lockObject)
+            {
+                if (running)
+                {
+                    cancellationTokenSource.Cancel();
+                    stopped.WaitOne();
+                    //clear the queue
+                    while (queue.Count > 0)
+                    {
+                        queue.Take();
+                    }
+                    cancellationTokenSource = null;
+                    running = false;
+                }
+            }
+        }
+
+        public void Enqueue(MethodInvocation methodInvocation)
+        {
+            var task = new Task<MethodInvocationResult>(() => ExecuteMethodInvocation(methodInvocation));
+            queue.Add(task);
+        }
+
+        private void ConsumeTasks()
+        {
+            try
+            {
+                while (!cancellationTokenSource.IsCancellationRequested)
+                {
+                    var task = queue.Take(cancellationTokenSource.Token);
+                    task.RunSynchronously(TaskScheduler.Current);
```

Comment:
Oh I missed that, I should've used the overload without a TaskScheduler parameter. So it will run on the consumer thread as expected.